### PR TITLE
feat(ir): add OpMemorySpaceSpec to op registry with auto-move insertion

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

- Add `OpMemorySpaceSpec` to the op registry, enabling declarative input/output memory constraints for all tile ops via fluent API (`set_input_memory`, `set_output_memory`, `set_output_memory_from_kwarg`, `set_output_memory_inherit_input`, `no_memory_spec`)
- Annotate all 80+ tile ops with memory space specs (Vec ops, Acc/cube ops, kwarg-based ops, view/inherit ops)
- Refactor `InferTileMemorySpace` pass to a 3-phase design (analyze → collect moves → mutate) that auto-inserts `tile.move` ops when input constraints are violated
- Remove hardcoded `tile.move` insertion from `ConvertTensorToTileOps` — memory space concerns are now centralized in `InferTileMemorySpace`
- Refactor `InitMemRef` to resolve memory spaces via registry specs instead of hardcoded op-name tables
- Add `ValidateTileOps()` import-time check: `import pypto` fails if any `tile.*` op is missing a memory spec
- Expose `get_op_memory_spec()` Python binding with type stubs

## Testing
- [x] All 2315 tests pass (no regressions)
- [x] 105 registry spec + infrastructure tests in `test_op_registry.py`
- [x] 25 pass behavior tests (inherit, auto-move, edge cases) in `test_infer_tile_memory_space.py`
- [x] 37 conversion tests updated for auto-move in `test_convert_tensor_to_tile_ops.py`
- [x] clang-tidy clean on changed files
- [x] pyright clean

### CODE REVIEW

File: python/bindings/modules/ir.cpp [hunk 1]
```diff
@@ -386,6 +386,40 @@ void BindIR(nb::module_& m) {
...
+        // Empty spec (from no_memory_spec()) — no constraints and no resolver
+        if (spec->input_constraints.empty() && !spec->deduce_output_memory) return nb::none();
+
+        nb::dict result;
+        // Input constraints
+        nb::list inputs;
+        for (const auto& c : spec->input_constraints) {
+          nb::list allowed;
+          for (auto ms : c) allowed.append(ms);
+          inputs.append(!allowed.empty() ? nb::object(nb::cast(allowed)) : nb::none());
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/bindings/modules/ir.cpp]: Fixed in 7b212c2 — unconstrained inputs now return an empty list instead of None, consistent with the C++ semantics and type stubs.

File: src/ir/transforms/infer_tile_memory_space_pass.cpp [hunk 2]
```diff
@@ -190,17 +243,114 @@ class TileMemorySpaceMutator : public IRMutator {
...
+
+    // Create moved var with memory_space_ set
+    auto move_type = As<TileType>(move_call->GetType());
+    INTERNAL_CHECK(move_type) << "Internal error: tile.move return type is not TileType";
+    auto moved_type = std::make_shared<TileType>(move_type->shape_, move_type->dtype_, move_type->memref_,
+                                                 move_type->tile_view_, target);
+    auto mutated_producer_var = As<Var>(mutated_producer);
+    INTERNAL_CHECK(mutated_producer_var) << "Internal error: mutated producer is not a Var";
+    auto moved_var = std::make_shared<Var>(mutated_producer_var->name_ + "_" + MemorySpaceToString(target),
+                                           std::move(moved_type), span);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/infer_tile_memory_space_pass.cpp:338]: Acknowledged. In practice, name collisions are unlikely since move variables are inserted in a scoped context and the naming pattern is predictable. If this becomes an issue, we can add a unique suffix counter. For now the current approach is simple and sufficient.

File: src/ir/transforms/init_memref.cpp [hunk 3]
```diff
@@ -43,29 +44,34 @@ namespace ir {
...
+
+  const auto& spec_opt = registry.GetEntry(op_name).GetMemorySpec();
+  if (!spec_opt.has_value() || !spec_opt->deduce_output_memory) {
+    return MemorySpace::Vec;
   }
-  // If target_memory not found, default to Vec
-  return MemorySpace::Vec;
+
+  auto result = spec_opt->deduce_output_memory(call->kwargs_);
+  return result.value_or(MemorySpace::Vec);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/init_memref.cpp:60]: Good observation. The Vec fallback for view ops in InitMemRef is harmless because the mutator shares the MemRef from the input rather than using the map value. Refactoring to use std::optional would be cleaner but is a separate concern — filed as a follow-up improvement.

[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/init_memref.cpp:60]: Good point. The unregistered cube-op fallback in InferTileMemorySpace is a backward-compat bridge for ops not yet registered. As we register more ops (like tile.matmul_mx*), this gap will close. For now InitMemRef's Vec fallback is safe because InferTileMemorySpace runs first and sets the correct TileType memory_space.

File: src/ir/transforms/infer_tile_memory_space_pass.cpp [hunk 4]
```diff
@@ -190,17 +243,114 @@ class TileMemorySpaceMutator : public IRMutator {
...
+
+      // After producer AssignStmt, insert tile.move stmts if needed
+      if (auto assign = As<AssignStmt>(stmt)) {
+        for (const auto& key : needed_moves_) {
+          if (key.first == assign->var_) {
+            InsertMoveStmt(new_stmts, assign->var_, key.second, assign->span_);
+            changed = true;
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/infer_tile_memory_space_pass.cpp:313]: Acknowledged. The needed_moves_ set is typically very small (1-3 entries per function), so the current O(stmts × moves) is practically O(stmts). If this becomes a bottleneck for large InCore functions, indexing by VarPtr would be a straightforward optimization.
